classes/local/report/location_report_builder.php

```diff
@@ -50,31 +50,52 @@ public function build_report($reportid) {
         $filedircount = 0;
         $filedirsum = 0;
         foreach ($locations as $location) {
-            $localsql = '';
+            $sql =
+                'WITH
+                  cte_objects AS (
+                    SELECT o.contenthash, o.location
+                      FROM {tool_objectfs_objects} o
+                     WHERE o.location = ? ),
+                  cte_obj_files AS (
+                    SELECT f.contenthash, MAX(f.filesize) AS filesize
+                      FROM {files} f
+                INNER JOIN cte_objects co ON f.contenthash = co.contenthash
+                     WHERE filesize > 0
+                  GROUP BY f.contenthash, f.filesize)
+               SELECT COALESCE(COUNT(cof.contenthash),0) AS objectcount,
+                      COALESCE(SUM(cof.filesize),0) AS objectsum
+                 FROM cte_obj_files cof';
+
             if ($location == OBJECT_LOCATION_LOCAL) {
-                $localsql = ' or o.location IS NULL';
+                $sql =
+                    'WITH
+                      cte_objects AS (
+                        SELECT o.contenthash,  o.location
+                          FROM {tool_objectfs_objects} o ),
+                      cte_obj_files AS (
+                        SELECT f.contenthash, MAX(f.filesize) AS filesize
+                          FROM {files} f
+                     LEFT JOIN cte_objects co ON f.contenthash = co.contenthash
+                         WHERE filesize > 0 AND ( co.location = ? OR co.location IS NULL )
+                      GROUP BY f.contenthash, f.filesize)
+                   SELECT COALESCE(COUNT(cof.contenthash),0) AS objectcount,
+                          COALESCE(SUM(cof.filesize),0) AS objectsum
```

Comment:
These SQL strings are largely similar to each other. Any way we can split them up into reusable chunks?

Reply:
It could probably be done.  I guess there's some balance in reusable chunks versus being more readable.

In the original query it was both reusable and readable, with the "OR LOCATION IS NULL" bit added conditionally.  But to make the query more optimized, that `WHERE` clause also shifts depending if we do the `LEFT JOIN` or `INNER JOIN` version.

I think if we tried to split them up into more reusable chunks it may be harder to understand the code and the reasoning why the query changes.  We could make it even more readable by doing a switch-case since it's really those 3 cases where the query is different: 

- for LOCAL we use the LEFT JOIN, and the WHERE LOCATION = 0 OR LOCATION = NULL clause comes after the LEFT JOIN;
- for DUPLICATED, EXTERNAL and ERROR we can do the INNER JOIN and we shift the WHERE LOCATION = ? clause to the `cte_objects` temp table, both improving performance;
- for ORPHANED we're only counting objects from `{tool_objectfs_objects}`.

I think it's okay as it is, but we could have a go at splitting up into more reusable chunks if you want?